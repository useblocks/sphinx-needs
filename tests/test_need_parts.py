import json
import os
from pathlib import Path

import pytest
from sphinx.util.console import strip_colors


@pytest.mark.parametrize(
    "test_app",
    [{"buildername": "html", "srcdir": "doc_test/doc_need_parts", "no_plantuml": True}],
    indirect=True,
)
def test_doc_need_parts(test_app, snapshot):
    app = test_app
    app.build()

    warnings = strip_colors(
        app._warning.getvalue().replace(str(app.srcdir) + os.sep, "srcdir/")
    ).splitlines()
    # print(warnings)
    assert warnings == []

    html = Path(app.outdir, "index.html").read_text()
    assert (
        '<span class="need-part" id="SP_TOO_001.1">exit()<a class="needs-id reference internal" '
        'href="#SP_TOO_001.1"> 1</a></span>' in html
    )

    assert '<em class="xref need">exit() (SP_TOO_001.1)</em>' in html
    assert '<em class="xref need">start() (SP_TOO_001.2)</em>' in html
    assert '<em class="xref need">blub() (SP_TOO_001.awesome_id)</em>' in html
    assert (
        '<em class="xref need">My custom link name (SP_TOO_001.awesome_id)</em>' in html
    )
    assert "SP_TOO_001" in html

    data = json.loads(Path(app.outdir, "needs.json").read_text())
    assert data["versions"][""]["needs"] == snapshot
