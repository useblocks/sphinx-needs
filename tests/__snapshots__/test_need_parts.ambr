# serializer version: 1
# name: test_doc_need_parts[test_app0]
  dict({
    'SP_TOO_001': dict({
      'content': '''
        The Tool awesome shall have a command line interface with following commands:
        
        * :need_part:`(1)exit()`
        * :need_part:`(2)start()`
        * :need_part:`(awesome_id)blub()`
        * :need_part:`(multiline_id)
          has multi-lines`
        
        
        * :np:`unknown_id_1`
        * :np:`unknown_id_2`
        
        .. spec:: TEST_2
            :id: TEST_2
        
            Part in nested need: :need_part:`(nested_id)something`
      ''',
      'docname': 'index',
      'external_css': 'external_link',
      'full_title': 'Command line interface',
      'id': 'SP_TOO_001',
      'layout': '',
      'lineno': 4,
      'parent_needs_back': list([
        'TEST_2',
      ]),
      'parts': dict({
        '1': dict({
          'content': 'exit()',
          'id': '1',
          'links': list([
          ]),
          'links_back': list([
          ]),
        }),
        '2': dict({
          'content': 'start()',
          'id': '2',
          'links': list([
          ]),
          'links_back': list([
          ]),
        }),
        '5DA': dict({
          'content': 'unknown_id_1',
          'id': '5DA',
          'links': list([
          ]),
          'links_back': list([
          ]),
        }),
        '7FF': dict({
          'content': 'unknown_id_2',
          'id': '7FF',
          'links': list([
          ]),
          'links_back': list([
          ]),
        }),
        'awesome_id': dict({
          'content': 'blub()',
          'id': 'awesome_id',
          'links': list([
          ]),
          'links_back': list([
          ]),
        }),
        'multiline_id': dict({
          'content': '''
            
            has multi-lines
          ''',
          'id': 'multiline_id',
          'links': list([
          ]),
          'links_back': list([
          ]),
        }),
      }),
      'section_name': 'NEED PARTS',
      'sections': list([
        'NEED PARTS',
      ]),
      'status': 'implemented',
      'tags': list([
        'test',
        'test2',
      ]),
      'title': 'Command line interface',
      'type': 'spec',
      'type_name': 'Specification',
    }),
    'TEST_2': dict({
      'content': 'Part in nested need: :need_part:`(nested_id)something`',
      'docname': 'index',
      'external_css': 'external_link',
      'full_title': 'TEST_2',
      'id': 'TEST_2',
      'layout': '',
      'lineno': 21,
      'parent_need': 'SP_TOO_001',
      'parent_needs': list([
        'SP_TOO_001',
      ]),
      'parts': dict({
        'nested_id': dict({
          'content': 'something',
          'id': 'nested_id',
          'links': list([
          ]),
          'links_back': list([
          ]),
        }),
      }),
      'section_name': 'NEED PARTS',
      'sections': list([
        'NEED PARTS',
      ]),
      'title': 'TEST_2',
      'type': 'spec',
      'type_name': 'Specification',
    }),
  })
# ---
